File: ruby/209_minimum_size_subarray_sum.rb
# @param {Integer} s
# @param {Integer[]} nums
# @return {Integer}

# Input: s = 7, nums = [2,3,1,2,4,3]
# Output: 2
# Explanation: the subarray [4,3] has the minimal length under the problem constraint.

# 一种依次遍历的方法，时间复杂度O(n^2), 超时了，所以得换种方法
# 不过通过这种方法的实验，知道了以下几种特殊情况
# 1. 数组为空
# 2. 数组和小于s
# 3. 数组中一个元素大于s

def min_sub_array_len(s, nums)
    return 0 if nums.empty? || nums.sum < s
    left, right = 0, -1
    res = Array.new
    sum = nums[left]
    while right < nums.size
        sum += nums[right] if right > left
        if sum >= s
            return 1 if right == -1
            res.push(right-left+1)
            left += 1
            sum = nums[left]
            right = -1
            next
        else
            right += 1
        end
    end
    res.min
end

# 另一种思路，我们一次遍历，每次累加sum，当sum >= s 时，此时说明当前的长度是符合要求的
# 然后将left往后移动，每次减小长度，最后返回最短的，此时时间复杂度为O(n)

def min_sub_array_len_version2(s, nums)
    return 0 if nums.empty? || nums.sum < s 
    
    sum, left = 0, 0
    res = Array.new
    for i in 0...nums.size
        sum += nums[i]
        while sum >= s 
            res.push(i+1-left)
            sum -= nums[left]
            left += 1
        end
    end

    res.min
end

nums = [1,1] #[2,3,1,2,4,3] #[1,4,4]
puts min_sub_array_len(3, nums)

File: cpp/209_minimum_size_subarray_sum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 给定数组和sum，要求返回最短的子数组和大于sum的长度；
// 1> two pinter
class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        if (nums.empty()) { return 0; }
		int left = 0, right = 0, sum = 0;
		int cnt = INT_MAX;
		while (right < nums.size()) {
			sum += nums[right++];
			while (sum >= s) {
				cnt = min(cnt, right-left);
				sum -= nums[left++];
			}
		}

		return cnt == INT_MAX ? 0 : cnt;
	}

	int bs(vector<int>& prefix, int sum, int left, int right) {
		while (left <= right) {
			int mid = left + (right - left) / 2;
			if (prefix[mid] >= sum) {
				if (mid == 0 || prefix[mid-1] < sum) {
					return mid;
				} else {
					right = mid - 1;
				}
			} else {
				left = mid + 1;
			}
		}
		return -1;
	}

	int minSubArrayLenBS(int s, vector<int>& nums) {
		int nsize = nums.size();
		
		vector<int> prefix(nsize+1, 0);
		for (int i = 0; i < nsize; ++i) {
			if (nums[i] >= s) { return 1; }
			prefix[i+1] = prefix[i] + nums[i];
		}

		int minLen = INT_MAX;
		for (int i = 0; i < nsize; ++i) {
			int pos = bs(prefix, prefix[i]+s, i+1, nsize);
			if (pos == -1) { continue; }
			minLen = min(minLen, pos - i);
		}
		
		return minLen == INT_MAX ? 0 : minLen;
	}
};
